src/app/tests/TestPowerSourceCluster.cpp

```diff
@@ -0,0 +1,266 @@
+
+    // OK, we should be in the right spot now, let's decode the list.
+    Clusters::PowerSource::Attributes::EndpointList::TypeInfo::DecodableType list;
+    err = list.Decode(attrDataReader);
+    NL_TEST_ASSERT(apSuite, err == CHIP_NO_ERROR);
+    std::vector<EndpointId> ret;
+    auto iter = list.begin();
+    while (iter.Next())
+    {
+        ret.push_back(iter.GetValue());
+    }
+    return ret;
+}
+
+void TestPowerSourceCluster::TestEndpointList(nlTestSuite * apSuite, void * apContext)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    Clusters::PowerSourceServer & powerSourceServer = Clusters::PowerSourceServer::Instance();
+
+    // test that when we read everything we get an empty list as nothing has been set up yet
+    for (EndpointId i = 0; i < 11; ++i)
+    {
+        std::vector<EndpointId> vec = ReadEndpointsThroughAttributeReader(apSuite, i);
+        NL_TEST_ASSERT(apSuite, vec.size() == 0);
+    }
+
+    // Test setting and getting
+    EndpointId list0[5]    = { 1, 2, 3, 4, 5 };
+    EndpointId list1[10]   = { 6, 7, 8, 9, 10, 11, 12, 13, 14, 15 };
+    EndpointId listRest[1] = { 2 };
+
+    err = powerSourceServer.SetEndpointList(0, Span<EndpointId>(list0));
+    NL_TEST_ASSERT(apSuite, err == CHIP_NO_ERROR);
+    const Span<EndpointId> * readBack = powerSourceServer.GetEndpointList(0);
+    NL_TEST_EXIT_ON_FAILED_ASSERT(apSuite, readBack != nullptr);
+    NL_TEST_ASSERT(apSuite, readBack->size() == 5);
+    for (size_t i = 0; i < readBack->size(); ++i)
+    {
+        NL_TEST_ASSERT(apSuite, readBack->data()[i] == list0[i]);
+    }
+
+    err = powerSourceServer.SetEndpointList(1, Span<EndpointId>(list1));
+    NL_TEST_ASSERT(apSuite, err == CHIP_NO_ERROR);
+    readBack = powerSourceServer.GetEndpointList(1);
+    NL_TEST_EXIT_ON_FAILED_ASSERT(apSuite, readBack != nullptr);
+    NL_TEST_ASSERT(apSuite, readBack->size() == 10);
+    for (size_t i = 0; i < readBack->size(); ++i)
+    {
+        NL_TEST_ASSERT(apSuite, readBack->data()[i] == list1[i]);
+    }
+
+    // Fill up the list
+    for (EndpointId i = 2; i < 10; ++i)
+    {
+        err = powerSourceServer.SetEndpointList(i, Span<EndpointId>(listRest));
+        NL_TEST_ASSERT(apSuite, err == CHIP_NO_ERROR);
+        readBack = powerSourceServer.GetEndpointList(i);
+        NL_TEST_EXIT_ON_FAILED_ASSERT(apSuite, readBack != nullptr);
+        NL_TEST_ASSERT(apSuite, readBack->size() == 1);
+        NL_TEST_ASSERT(apSuite, readBack->data()[0] == listRest[0]);
+    }
+
+    // Add one more, make sure we get an out of memory error
+    err = powerSourceServer.SetEndpointList(11, Span<EndpointId>(listRest));
+    NL_TEST_ASSERT(apSuite, err == CHIP_ERROR_NO_MEMORY);
+
+    // Make sure all the stuff we had before is still there
+    readBack = powerSourceServer.GetEndpointList(0);
+    NL_TEST_EXIT_ON_FAILED_ASSERT(apSuite, readBack != nullptr);
+    NL_TEST_ASSERT(apSuite, readBack->size() == 5);
+    for (size_t i = 0; i < readBack->size(); ++i)
```

Comment:
Recomend naming the "i" to what it is like "endpointIdx" above, or "listIdx" here, etc.